### PR TITLE
Don't inherit Shape from NamedTuple

### PR DESCRIPTION
**Pull request by awygle**
Fixes issue #421 .

### CODE REVIEW

File: nmigen/hdl/ast.py [hunk 1]
```diff
@@ -55,8 +71,15 @@ class Shape(typing.NamedTuple):
...
-    signed: bool = False
+    def __init__(self, width=1, signed=False):
+        if not isinstance(width, int) or width < 0:
+            raise TypeError("Width must be a non-negative integer, not {!r}"
+                            .format(width))
+        self.width = width
+        self.signed = signed
+    
+    def __iter__(self):
+        return ShapeIter(self.width, self.signed)
```
Comment (whitequark) [inline nmigen/hdl/ast.py:82]: Why not `return iter((self.width, self.signed))`?

**Comment by awygle** [comment]
Because I'm bad at python and didn't know iter() existed, and also I've
been writing Rust a lot and this is how you have to do it there.

File: nmigen/test/test_hdl_ast.py [hunk 2]
```diff
@@ -39,6 +39,11 @@ def test_make_wrong(self):
                 msg="Width must be a non-negative integer, not -1"):
             Shape(-1)
 
+    def test_compare_wrong(self):
+        with self.assertRaises(TypeError,
+                msg="Shapes may be compared with other Shapes and (int, bool) tuples, not (2, 3)"):
+            Shape(1, True) == (2, 3)
```
Comment (whitequark) [inline nmigen/test/test_hdl_ast.py:45]: Needs a test for the second `raise TypeError`.

**Comment by whitequark** [comment]
Thanks!

**Pull request merged**